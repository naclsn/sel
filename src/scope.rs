use std::collections::HashMap;
use std::fs::File;
use std::io::{Read, Result as IoResult};
use std::mem;
use std::ops::Range;
use std::path::{Path, PathBuf};
use std::ptr;

use crate::builtin;
use crate::parse::Tree;
use crate::types::{TypeList, TypeRef};

pub struct Global {
    pub registry: SourceRegistry,
    pub types: TypeList,
    pub scope: Scope,
}

#[derive(Debug)]
pub enum ScopeItem {
    Builtin(fn(&mut TypeList) -> TypeRef, &'static str),
    Defined(Tree, String),
    Binding(Location, TypeRef),
}

#[derive(Debug)]
pub struct Scope {
    parent: *const Scope,
    names: HashMap<String, ScopeItem>,
}

// source registry {{{
pub type SourceRef = usize;

#[derive(PartialEq, Debug, Clone)]
pub struct Location(pub SourceRef, pub Range<usize>);

#[derive(Default)]
pub struct SourceRegistry(Vec<(PathBuf, Option<Vec<u8>>)>);

impl SourceRegistry {
    /// Note: the given path is taken as is (ie not canonicalize) and not checked for duplication
    pub fn add_bytes(&mut self, name: impl AsRef<Path>, bytes: Vec<u8>) -> SourceRef {
        self.0.push((name.as_ref().to_owned(), Some(bytes)));
        self.0.len() - 1
    }

    pub fn add(&mut self, name: impl AsRef<Path>) -> IoResult<SourceRef> {
        let name = name.as_ref().canonicalize()?;
        if let Some(found) = self.0.iter().position(|c| c.0 == name) {
            return Ok(found);
        }
        let mut bytes = Vec::new();
        File::open(&name)?.read_to_end(&mut bytes)?;
        Ok(self.add_bytes(name, bytes))
    }

    pub fn get_path(&self, at: SourceRef) -> &Path {
        &self.0[at].0
    }

    pub fn get_bytes(&self, at: SourceRef) -> &[u8] {
        self.0[at].1.as_ref().unwrap()
    }

    pub fn take_bytes(&mut self, at: SourceRef) -> Vec<u8> {
        self.0[at].1.take().unwrap()
    }

    pub fn put_back_bytes(&mut self, at: SourceRef, bytes: Vec<u8>) {
        self.0[at].1 = Some(bytes);
    }
}
// }}}

impl Global {
    pub fn with_builtin() -> Global {
        Global {
            registry: SourceRegistry::default(),
            types: TypeList::default(),
            scope: builtin::scope(),
        }
    }
}

// scope {{{
impl ScopeItem {
    pub fn make_type(&self, types: &mut TypeList) -> TypeRef {
        match self {
            ScopeItem::Builtin(mkty, _) => mkty(types),
            ScopeItem::Defined(val, _) => types.duplicate(val.ty, &mut HashMap::new()),
            ScopeItem::Binding(_, ty) => types.duplicate(*ty, &mut HashMap::new()),
        }
    }

    pub fn get_loc(&self) -> Option<&Location> {
        match self {
            ScopeItem::Builtin(_, _) => None,
            ScopeItem::Defined(val, _) => Some(&val.loc),
            ScopeItem::Binding(loc, _) => Some(loc),
        }
    }

    pub fn get_desc(&self) -> Option<&str> {
        match self {
            ScopeItem::Builtin(_, desc) => Some(desc),
            ScopeItem::Defined(_, desc) => Some(desc),
            ScopeItem::Binding(_, _) => None,
        }
    }
}

// YYY: weird pattern, surely there is better :/
pub(crate) struct MustRestore(Scope);

impl Default for Scope {
    fn default() -> Scope {
        Scope {
            parent: ptr::null(),
            names: HashMap::default(),
        }
    }
}

impl Scope {
    pub(crate) fn new(parent: &Scope) -> Scope {
        Scope {
            parent: parent as *const Scope,
            names: HashMap::new(),
        }
    }

    /// returns the parent scope (previously `self`)
    /// which is expected to be restored with `restore_from_parent`
    pub(crate) fn make_into_child(&mut self) -> MustRestore {
        let parent = mem::take(self);
        self.parent = &parent as *const Scope;
        MustRestore(parent)
    }
    /// see `make_into_child`
    pub(crate) fn restore_from_parent(&mut self, parent: MustRestore) {
        *self = parent.0;
    }

    /// parent-most englobing scope
    pub fn global(&self) -> &Scope {
        unsafe { self.parent.as_ref() }.unwrap_or(self)
    }

    /// does not insert if the name already existed, returns a reference to the previous value
    pub(crate) fn declare(&mut self, name: String, value: ScopeItem) -> Option<&ScopeItem> {
        let mut already = false;
        let r = self
            .names
            .entry(name)
            .and_modify(|_| already = true)
            .or_insert_with(|| value);
        if already {
            Some(r)
        } else {
            None
        }
    }

    pub fn lookup(&self, name: &str) -> Option<&ScopeItem> {
        self.names
            .get(name)
            .or_else(|| unsafe { self.parent.as_ref() }?.lookup(name))
    }

    pub fn iter(&self) -> <&HashMap<String, ScopeItem> as IntoIterator>::IntoIter {
        self.names.iter()
    }

    pub fn iter_rec(&self) -> ScopeRecIter {
        ScopeRecIter {
            entries: self.names.iter(),
            parent: unsafe { self.parent.as_ref() },
        }
    }
}

// iters {{{
pub struct ScopeRecIter<'a> {
    entries: <&'a HashMap<String, ScopeItem> as IntoIterator>::IntoIter,
    parent: Option<&'a Scope>,
}

impl<'a> Iterator for ScopeRecIter<'a> {
    type Item = (&'a String, &'a ScopeItem);

    fn next(&mut self) -> Option<Self::Item> {
        self.entries.next().or_else(|| {
            *self = self.parent?.iter_rec();
            self.next()
        })
    }
}

impl IntoIterator for Scope {
    type IntoIter = <HashMap<String, ScopeItem> as IntoIterator>::IntoIter;
    type Item = <Self::IntoIter as Iterator>::Item;

    fn into_iter(self) -> Self::IntoIter {
        self.names.into_iter()
    }
}

impl<'a> IntoIterator for &'a Scope {
    type IntoIter = <&'a HashMap<String, ScopeItem> as IntoIterator>::IntoIter;
    type Item = <Self::IntoIter as Iterator>::Item;

    fn into_iter(self) -> Self::IntoIter {
        self.names.iter()
    }
}
// }}}
// }}}
