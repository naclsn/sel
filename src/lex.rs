//! lexing

use std::iter::{self, Enumerate, Peekable};

use crate::scope::{Location, SourceRef};

#[derive(PartialEq, Debug, Clone)]
pub enum TokenKind {
    Unknown(String),
    Number(f64),
    Bytes(Box<[u8]>),
    Word(String),
    Comma,
    OpenBracket,
    CloseBracket,
    OpenBrace,
    CloseBrace,
    Equal,
    Def,
    Let,
    Use,
    //LineComment(String),
    //DashComment(String),
    End,
}

#[derive(PartialEq, Debug, Clone)]
pub struct Token(pub Location, pub TokenKind);

/// note: this is an infinite iterator (`next()` is never `None`)
pub struct Lexer<I: Iterator<Item = u8>> {
    stream: Peekable<Enumerate<I>>,
    source: SourceRef,
    last_at: usize,
}

impl<I: Iterator<Item = u8>> Lexer<I> {
    pub(crate) fn new(source: SourceRef, iter: I) -> Self {
        Self {
            stream: iter.enumerate().peekable(),
            source,
            last_at: 0,
        }
    }
}

impl<I: Iterator<Item = u8>> Iterator for Lexer<I> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        use TokenKind::*;

        let Some((at, byte)) = self.stream.find(|c| !c.1.is_ascii_whitespace()) else {
            let range = self.last_at..self.last_at;
            return Some(Token(Location(self.source, range), End));
        };

        let (len, tok) = match byte {
            b':' => {
                let mut doubles = 2;
                let b: Box<[u8]> =
                    iter::from_fn(|| match (self.stream.next(), self.stream.peek()) {
                        (Some((_, b':')), Some((_, b':'))) => {
                            doubles += 1;
                            self.stream.next();
                            Some(b':')
                        }
                        (Some((_, b':')), _) => None,
                        (pair, _) => pair.map(|c| c.1),
                    })
                    .collect();
                (b.len() + doubles, Bytes(b))
            }

            b',' => (1, Comma),
            b'[' => (1, OpenBracket),
            b']' => (1, CloseBracket),
            b'{' => (1, OpenBrace),
            b'}' => (1, CloseBrace),
            b'=' => (1, Equal),

            b'#' => {
                if let Some((_, b'-')) = self.stream.peek() {
                    self.stream.next();
                    let mut m = Vec::new();
                    while {
                        match self.next().expect("unreachable: infinite iterator").1 {
                            OpenBracket => m.push(b'b'),
                            CloseBracket if m.last().is_some_and(|b| b'b' == *b) => _ = m.pop(),
                            OpenBrace => m.push(b'B'),
                            CloseBrace if m.last().is_some_and(|b| b'B' == *b) => _ = m.pop(),
                            End => m.clear(),
                            _ => {}
                        }
                        !m.is_empty()
                    } {}
                    let n = self.next();
                    let Some(Token(_, Comma)) = n else {
                        return n;
                    };
                } else {
                    self.stream.find(|c| {
                        self.last_at = c.0;
                        b'\n' == c.1
                    });
                }
                return self.next();
            }
            b'_' => (1, Word("_".into())),

            c if c.is_ascii_lowercase() || b'-' == c => {
                let b: Vec<u8> = iter::once(c)
                    .chain(iter::from_fn(|| {
                        self.stream
                            .next_if(|c| c.1.is_ascii_lowercase() || b'-' == c.1)
                            .map(|c| c.1)
                    }))
                    .collect();
                (
                    b.len(),
                    match &b[..] {
                        b"def" => Def,
                        b"let" => Let,
                        b"use" => Use,
                        _ => Word(String::from_utf8(b).expect("unreachable: is_ascii_lowercase")),
                    },
                )
            }

            c if c.is_ascii_digit() => {
                let mut r = 0;
                let mut l = 1;

                let (shift, digits) = match (c, self.stream.peek()) {
                    (b'0', Some((_, b'B' | b'b'))) => (1, b"01".as_slice()),
                    (b'0', Some((_, b'O' | b'o'))) => (3, b"01234567".as_slice()),
                    (b'0', Some((_, b'X' | b'x'))) => (4, b"0123456789abcdef".as_slice()),
                    _ => (0, b"0123456789".as_slice()),
                };
                if 0 == shift {
                    r = c as usize - 48;
                } else {
                    l += 1;
                    self.stream.next();
                }

                while let Some(k) = self
                    .stream
                    .peek()
                    .and_then(|c| digits.iter().position(|&k| k == c.1 | 32))
                {
                    l += 1;
                    self.stream.next();
                    r = if 0 == shift {
                        r * 10 + k
                    } else {
                        (r << shift) | k
                    };
                }

                let r = if 0 == shift && self.stream.peek().is_some_and(|c| b'.' == c.1) {
                    l += 2;
                    self.stream.next();
                    let mut d = match self.stream.next() {
                        Some(c) if c.1.is_ascii_digit() => (c.1 - b'0') as usize,
                        _ => {
                            self.last_at = at + l;
                            let t = r.to_string() + ".";
                            return Some(Token(Location(self.source, at..at + l), Unknown(t)));
                        }
                    };
                    let mut w = 10;
                    while let Some(c) = self.stream.peek().map(|c| c.1).filter(u8::is_ascii_digit) {
                        l += 1;
                        self.stream.next();
                        d = d * 10 + (c - b'0') as usize;
                        w *= 10;
                    }
                    r as f64 + (d as f64 / w as f64)
                } else {
                    r as f64
                };

                (l, Number(r))
            }

            c => {
                let chclass = c.is_ascii_alphanumeric();
                let b: Vec<u8> = iter::once(c)
                    .chain(iter::from_fn(|| {
                        self.stream
                            .next_if(|c| {
                                c.1.is_ascii_alphanumeric() == chclass && !c.1.is_ascii_whitespace()
                            })
                            .map(|c| c.1)
                    }))
                    .collect();
                (b.len(), Unknown(String::from_utf8_lossy(&b).into_owned()))
            }
        };

        self.last_at = at + len;
        Some(Token(Location(self.source, at..at + len), tok))
    }
}

#[test]
fn test() {
    use insta::assert_snapshot;

    fn t(script: &[u8]) -> String {
        let mut end = true;
        Lexer::new(0, script.iter().copied())
            .take_while(|t| {
                if !end {
                    false
                } else {
                    end = TokenKind::End != t.1;
                    true
                }
            })
            .fold(String::new(), |mut acc, cur| {
                acc += &format!("{cur:?}\n");
                acc
            })
    }

    assert_snapshot!(t(b""));
    assert_snapshot!(t(b"coucou"));
    assert_snapshot!(t(b"[a 1 b, {w, x, y, z}, 0.5]"));
    assert_snapshot!(t(b":hay: :hey:: not hay: :: :::: fin"));
    assert_snapshot!(t(b"42 :*: 0x2a 0b101010 0o52"));
    assert_snapshot!(t(b"oh-no; don't let use be the def of cat!"));
    assert_snapshot!(t(b"
normal
# comment
normal again
#-comment not comment
#- comment again not comment
a #- [ba ba ba] b
a #- 239084 b
a #- :yeeesldkfj: b
a #- {one, [two], [{three}]} b
a #- {one, #-[two], [{three}]} b
{heyo, #-baba, owieur} # the ',' is also commented
{heyo, baba, #-owieur} # the '}' isn't
"));
}
