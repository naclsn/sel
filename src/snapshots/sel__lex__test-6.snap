---
source: src/lex.rs
expression: "t(b\"oh-no; don't let use be the def of cat!\")"
---
Token(Location("<test>", 0..5), Word("oh-no"))
Token(Location("<test>", 5..6), Unknown(";"))
Token(Location("<test>", 7..10), Word("don"))
Token(Location("<test>", 10..11), Unknown("'"))
Token(Location("<test>", 11..12), Word("t"))
Token(Location("<test>", 13..16), Let)
Token(Location("<test>", 17..20), Use)
Token(Location("<test>", 21..23), Word("be"))
Token(Location("<test>", 24..27), Word("the"))
Token(Location("<test>", 28..31), Def)
Token(Location("<test>", 32..34), Word("of"))
Token(Location("<test>", 35..38), Word("cat"))
Token(Location("<test>", 38..39), Unknown("!"))
Token(Location("<test>", 39..39), End)
